src/main/java/gregtech/api/capability/impl/FluidDrillLogic.java

```diff
@@ -0,0 +1,317 @@
+package gregtech.api.capability.impl;
+
+import gregtech.api.GTValues;
+import gregtech.api.capability.GregtechDataCodes;
+import gregtech.api.metatileentity.MetaTileEntity;
+import gregtech.api.worldgen.bedrockFluids.BedrockFluidVeinHandler;
+import gregtech.common.ConfigHolder;
+import gregtech.common.metatileentities.multi.electric.MetaTileEntityFluidDrill;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidStack;
+
+import javax.annotation.Nonnull;
+
+public class FluidDrillLogic {
+
+    public static final int MAX_PROGRESS = 20;
+
+    private int progressTime = 0;
+
+    private final MetaTileEntityFluidDrill metaTileEntity;
+
+    private boolean isActive;
+    private boolean isWorkingEnabled = true;
+    private boolean wasActiveAndNeedsUpdate;
+    private boolean isDone = false;
+    protected boolean isInventoryFull;
+
+    private boolean hasNotEnoughEnergy;
+
+    private Fluid veinFluid;
+
+    public FluidDrillLogic(MetaTileEntityFluidDrill metaTileEntity) {
+        this.metaTileEntity = metaTileEntity;
+        this.veinFluid = null;
+    }
+
+    /**
+     * Performs the actual drilling
+     * Call this method every tick in update
+     */
+    public void performDrilling() {
+        if (metaTileEntity.getWorld().isRemote) return;
+
+        // if we have no fluid, try to get a new one
+        if (veinFluid == null)
+            if (!acquireNewFluid())
+                return; // stop if we still have no fluid
+
+
+        // drills that cannot work do nothing
+        if (!this.isWorkingEnabled)
+            return;
+
+        // check if drilling is possible
+        if (!checkCanDrain())
+            return;
+
+        // if the inventory is not full, drain energy etc. from the drill
+        // the storages have already been checked earlier
+        if (!isInventoryFull) {
+            // actually drain the energy
+            consumeEnergy(false);
+
+            // since energy is being consumed the rig is now active
+            if (!this.isActive)
+                setActive(true);
+        } else {
+            // the rig cannot drain, therefore it is inactive
+            if (this.isActive)
+                setActive(false);
+            return;
+        }
+
+        // increase progress
+        progressTime++;
+        if (progressTime % MAX_PROGRESS != 0)
+            return;
+        progressTime = 1;
```

Comment:
Shouldn't this reset to 0 then? As if it resets to 1, it will only take 19 ticks for a new operation to start, instead of the 1 second break like you wanted

Reply:
Resetting to 0 is actually 21 ticks of time. On tick 20/20, 20 ticks have passed if you start with ticks at 1. This is also how the abstract recipe logic does it.

Reply:
No, abstract recipe logic resets progress time to 0 upon completing a recipe.

Resetting to 0 here would be 20 ticks of time, because you are incrementing the progress time before checking the modulus. Therefore, on the first tick after reset, it would be: progressTime = 0 -> progress time = 1 -> check modulus.

How it is now is: progressTime = 1 -> progressTime = 2 -> check modulus. 
With this pattern, you will only need to run through `update()` 19 times before the modulus succeeds.